src/ape/pytest/contextmanagers.py

```diff
@@ -1,26 +1,151 @@
+            ) from exc
+
+        if trace[-1].op != "REVERT":
+            raise AssertionError("Transaction trace did not end with REVERT.")
+
+        pc = None
+        pcmap = contract.contract_type.pcmap.parse()
+
+        # To find a suitable line for inspecting dev messages, we must start at the revert and work
+        # our way backwards. If the last frame's PC is in the PC map, the offending line is very
+        # likely a 'raise' statement.
+        if trace[-1].pc in pcmap:
+            pc = trace[-1].pc
+
+        # Otherwise we must traverse the trace backwards until we find our first suitable candidate.
+        # It should be noted that when an error happens, typically the contract will execute:
+        #
+        #   <...>
+        #   JUMPI    <- jump to revert routine on a failing condition
+        #
+        #   <...>
+        #   JUMPDEST <- beginning of revert routine
+        #   PUSH1
+        #   DUP1
+        #   REVERT
+        #
+        # So we must step backwards until we find the initial instructions that caused the failing
+        # condition. This will not be the JUMPI or even an arithmetic instruction like DIV, but
+        # rather the starting instruction for a given line.
+        else:
+            while len(trace) > 0:
+                frame = trace.pop()
+                if frame.pc in pcmap:
+                    pc = frame.pc
+                    break
+
+        # We were unable to find a suitable PC that matched the compiler's map.
+        if pc is None:
+            raise AssertionError("Could not find line that caused revert.")
+
+        offending_source = pcmap[pc]
+
+        # The compiler PC map had PC information, but not source information.
+        if offending_source is None or offending_source.line_start is None:
+            raise AssertionError("Could not find line that caused revert.")
+
+        sources = ape.project.extract_manifest().sources or {}
+        source_id = contract.contract_type.source_id or ""
+
+        if not (sources and source_id):
+            AssertionError("Could not retrieve contract source.")
+
+        if source_id not in sources:
+            AssertionError("Could not retrieve contract source.")
+
+        content = sources[source_id].content or ""
+
+        if not content:
+            AssertionError("Could not retrieve contract source.")
 
-        actual = exc_value.revert_message
+        contract_source_lines = content.split("\n")
+
+        if offending_source.line_start > len(contract_source_lines):
+            raise AssertionError(
+                "Could not find line that caused revert (line outside of contract range)."
+            )
+
+        # Line numbers will not be zero-indexed
+        offending_line = contract_source_lines[offending_source.line_start - 1]
+
+        # Match '# dev: (...)' with any variation of whitespace
+        match = re.search(r"#\s*dev\s*:\s*(.+)", offending_line)
```

Comment:
This part seems tricky because we don't know anything about the language, like does it use `#` or `//` for comments or something entirely different?

We need to delegate some of this to the compiler plugins?

Reply:
Yeah good point

I think maybe for now since this really only works with Vyper, we could stick with this for the alpha

Then as part of APE-298/299 there could be a step to break some of this stuff out and delegate it to the plugins. I think that breaking this out would also kind of tie in with https://github.com/ApeWorX/ape/pull/1084#discussion_r1012180056 as well

Perhaps that "breaking out" step could be its own thing done before APE-298/299